src/Features/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/Testing/RunTestsHandler.cs

```diff
@@ -0,0 +1,131 @@
+            TraceLevel = GetTraceLevel(serverConfiguration),
+            EnvironmentVariables = new()
+            {
+                // Reset dotnet root to the user's original configuration so that vs test console can find the right runtimes.
+                { DotnetCliHelper.DotnetRootEnvVar, DotnetCliHelper.GetUserDotnetRoot() },
+            }
+        });
+
+        var testCases = await testDiscoverer.DiscoverTestsAsync(request.Range, context.Document, projectOutputPath, progress, vsTestConsoleWrapper, cancellationToken);
+        if (testCases.IsEmpty)
+        {
+            return progress.GetValues() ?? Array.Empty<RunTestsPartialResult>();
+        }
+
+        await testRunner.RunTestsAsync(testCases, progress, vsTestConsoleWrapper, cancellationToken);
+
+        return progress.GetValues() ?? Array.Empty<RunTestsPartialResult>();
+    }
+
+    /// <summary>
+    /// Shells out to the .NET CLI to build the project to ensure that its relatively up to date as of this request.
+    /// We can never be entirely sure that this build happens on exactly the same snapshot that this request was made on
+    /// as something could have changed on the client side immediately after this request was triggered.
+    ///
+    /// However if we don't do a build it is likely the user code is very different from what they last built
+    /// which results in a confusing experience.
+    /// </summary>
+    private async Task BuildAsync(Document document, BufferedProgress<RunTestsPartialResult> progress, CancellationToken cancellationToken)
+    {
+        var workingDirectory = Path.GetDirectoryName(document.Project.FilePath);
+        Contract.ThrowIfNull(workingDirectory, $"Unable to get working directory for project {document.Project.Name}");
+
+        // TODO - we likely need to pass the no-restore flag once we have automatic restore enabled.
+        // https://github.com/dotnet/vscode-csharp/issues/5725
+        var arguments = "build";
+        using var process = dotnetCliHelper.Run(arguments, workingDirectory);
+
+        process.OutputDataReceived += (sender, args) =>
+        {
+            if (args.Data != null)
+            {
+                ReportProgress(progress, args.Data);
+            }
+        };
+
+        process.ErrorDataReceived += (sender, args) =>
+        {
+            if (args.Data != null)
+            {
+                ReportProgress(progress, args.Data);
+            }
+        };
+        process.BeginOutputReadLine();
+        process.BeginErrorReadLine();
+
+        await process.WaitForExitAsync(cancellationToken);
+
+        if (process.ExitCode != 0)
+        {
+            throw new InvalidOperationException("Failed to run build, see test output for details");
+        }
+
+        static void ReportProgress(BufferedProgress<RunTestsPartialResult> progress, string buildOutput)
+        {
+            progress.Report(new RunTestsPartialResult("Building project...", buildOutput, Progress: null));
```

Comment:
are these strings that hte user will see?  should they be localized?

Reply:
I'm tracking localization in a followup PR.  I need to do some additional changes to fix localization in this project, so to avoid making this PR even larger I'm doing them in the next PR.